### DISCUSSION TITLE
How do you open App from an App Link

### DISCUSSION THREAD

**Discussion by zburlott4** [airsdk/Adobe-Runtime-Support#3269] [General]
Is is possible to open an android and/or apple version of an app using an app link. If it is, is there an example of how this is done. Use case is you send a link via sms or email and user clicks the link and it automatically prompts to open the app or immediately opens the app.

**Comment by ajwfrost** [comment]
I believe so .. was just looking at #2845 which says it can be done on Android/iOS but not on Windows (at least not without hooking up additional tools/installers etc).

For Android and iOS, I believe it would need to be added via the manifest / info.plist files respectively, using the `android.manifestAdditions` and `iPhone.InfoAdditions` sections of the app descriptor files. And then following platform-specific guidelines:

For Android:
https://stackoverflow.com/questions/2448213/how-to-implement-my-very-own-uri-scheme-on-android

For iPhone:
They suggest using Xcode of course, but ultimately the values find their way into Info.plist, and the below link appears to have the xml codes that you'd need in InfoAdditions:
https://stackoverflow.com/questions/8201724/how-to-register-a-custom-app-opening-url-scheme-with-xcode-4

Hope that helps!

**Comment by marchbold** [comment]
There are two methods for this:
- using a custom url scheme (sometimes referred to as deep links)
- using an app/universal link

## Custom URL Scheme

A custom url scheme involves a url like `myapp://somelocation` and you need to register the url scheme in your application descriptor. This method is an older approach but simpler, and there is no real method to handle the cases where the app isn't installed.

For iOS:

```xml
<key>CFBundleURLTypes</key>
<array>
    <dict>
        <key>CFBundleURLSchemes</key>
        <array>
            <string>myapp</string>
        </array>
    </dict>
</array>
```

Android:
```xml
<activity>
    <intent-filter>
        <action android:name="android.intent.action.MAIN"></action>
        <category android:name="android.intent.category.LAUNCHER"></category>
    </intent-filter>
    <intent-filter>
        <action android:name="android.intent.action.VIEW"/>
        <category android:name="android.intent.category.DEFAULT"/>
        <category android:name="android.intent.category.BROWSABLE"/>

        <data android:scheme="myapp"/>
    </intent-filter>
</activity>
```

> Note: This activity must be inside your `application` tag in your manifest additions. You can only have one activity without a name which will be used as the main activity by AIR. You must keep the first intent filter as is, the second is the additional one for the url scheme. 


## App / Universal Links

The more modern method is to use app/universal links. 

These are slightly more complex as you need to host content on a website which contains certain data and are specific to a platform. For both of these methods I suggest you follow the official documentation to create and verify the appropriate files on your server.

### Android App Links

The general steps for creating Android App Links are as follows:

1. Create deep links to specific content in your app: In your app manifest, create intent filters for your website URIs and configure your app to use data from the intents to send users to the right content in your app. Learn more in Create Deep Links to App Content. eg

```xml
<activity>
    <intent-filter>
        <action android:name="android.intent.action.MAIN"></action>
        <category android:name="android.intent.category.LAUNCHER"></category>
    </intent-filter>
    <intent-filter>
        <action android:name="android.intent.action.VIEW"/>
        <category android:name="android.intent.category.DEFAULT"/>
        <category android:name="android.intent.category.BROWSABLE"/>

        <data android:scheme="http" />
        <data android:scheme="https" />

        <data android:host="myappdomain.com" />
    </intent-filter>
</activity>
```

2. Add verification for your deep links: Configure your app to request verification of app links. Then, publish a Digital Asset Links JSON file on your websites to verify ownership through Google Search Console. Learn more in [Verify App Links](https://developer.android.com/training/app-links/verify-android-applinks). An example `assetlinks.json` file:

```json
[{
  "relation": ["delegate_permission/common.handle_all_urls"],
  "target": {
    "namespace": "android_app",
    "package_name": "com.myapp",
    "sha256_cert_fingerprints":
    ["14:6D:E9:83:C5:73:06:50:D8:EE:B9:95:2F:34:FC:64:16:A0:83:42:E6:1D:BE:A8:8A:04:96:B2:3F:CF:44:E5"]
  }
}]
```

You can then manage and verify app links through the play console.


### iOS Universal Links

1. Create an "association file" and connect your associated domain: 

https://developer.apple.com/documentation/xcode/supporting-associated-domains?language=objc

Example `apple-app-site-association` file:

```json
{
  "applinks": {
    "apps": [],
    "details": [
      {
        "appID": "ABCDE12345.com.myapp",
        "paths": [
          "/",
          "/product/*",
          "*.html"
        ]
      }
    ]
  },
  "webcredentials": {
    "apps": ["ABCDE12345.com.myapp"]
  },
  "appclips": {
    "apps": ["ABCDE12345.com.myapp"]
  }
}
```

2. Add the domains to your application's entitlements. 

```xml
<key>com.apple.developer.associated-domains</key>
<array>
    <string>applinks:myappdomain.com</string>
</array>
```
